manifests/pip.pp

```diff
@@ -215,14 +215,21 @@
         }
 
         # Unfortunately this is the smartest way of getting the latest available package version with pip as of now
-        # Note: we DO need to repeat ourselves with "from version" in both grep and sed as on some systems pip returns
-        # more than one line with paretheses.
-        $latest_version = join([
-            "${pip_install} ${legacy_resolver} ${pypi_index} ${pypi_extra_index} ${proxy_flag}",
-            " ${install_args} ${install_editable} '${real_pkgname}==9!0dev0+x' 2>&1",
-            " | sed -nE 's/.*\\(from versions: (.*, )*(.*)\\)/\\2/p'",
-            ' | tr -d "[:space:]"',
-        ])
+        if $install_args and $install_args =~ /--pre/ {
+          $latest_version = join([
+              "${pip_install} ${legacy_resolver} ${pypi_index} ${pypi_extra_index} ${proxy_flag}",
+              " ${install_args} ${install_editable} ${real_pkgname}==9!0dev0+x 2>&1",
+              " | sed -nE 's/.*\\(from versions: (.*, )*(.*)\\)/\\2/p'",
+              ' | tr -d "[:space:]"',
+          ])
+        } else {
+          $latest_version = join([
+              "${pip_install} ${legacy_resolver} ${pypi_index} ${pypi_extra_index} ${proxy_flag}",
+              " ${install_args} ${install_editable} ${real_pkgname}==9!0dev0+x 2>&1",
+              " | sed -nE 's/.*\\(from versions: ([^\\)]*)\\)/\\1/p'",
+              ' | awk \'BEGIN {RS=", "} {if ($0 !~ /(a|b|rc|dev)/) {gsub(/\n/,"");stable[arraylen++]=$0}} END {print stable[arraylen-1] }\'',
+          ])
+        }
```

Comment:
It would be better to avoid duplicating the first two lines of the command, like this:

```puppet
        $version_processing_commands = if $install_args and $install_args =~ /--pre/ {
          [
            " | sed -nE 's/.*\\(from versions: (.*, )*(.*)\\)/\\2/p'",
            ' | tr -d "[:space:]"',
          ]
        } else {
          [
            " | sed -nE 's/.*\\(from versions: ([^\\)]*)\\)/\\1/p'",
            ' | awk \'BEGIN {RS=", "} {if ($0 !~ /(a|b|rc|dev)/) {gsub(/\n/,"");stable[arraylen++]=$0}} END {print stable[arraylen-1] }\'',
          ]
        }

        $latest_version = join(
          [
            "${pip_install} ${legacy_resolver} ${pypi_index} ${pypi_extra_index} ${proxy_flag}",
            " ${install_args} ${install_editable} '${real_pkgname}==9!0dev0+x' 2>&1",
          ] + $version_processing_commands,
        )
```

I think we want to keep the single quotation marks around `${real_pkgname}==9!0dev0+x` too, right?